include/sys/eventfd.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2020 Tobias Svehagen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_EVENTFD_H_
+#define ZEPHYR_INCLUDE_SYS_EVENTFD_H_
+
+#include <sys/fdtable.h>
+#include <sys/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef u32_t eventfd_t;
+
+/**
+ * @brief Create a file descriptor for event notification
+ *
+ * The returned file descriptor can be used with POSIX read/write calls or
+ * with the eventfd_read/eventfd_write functions.
+ *
+ * It also supports polling and by including an eventfd in a call to poll,
+ * it is possible to signal and wake the polling thread by simply writing to
+ * the eventfd.
+ *
+ * When using read() and write() on an eventfd, the size must always be at
+ * least 4 bytes or the operation will fail with EINVAL.
+ *
+ * @return New eventfd file descriptor on success, -1 on error
+ */
+int eventfd(void);
```

Comment:
> I would think it would be better to not have functions in Zephyr with the same name as other functions in known APIs, almost matching them but not quite.

Yes, this is a good idea in order to avoid confusion later.

Reply:
Any suggestions for name? Another option I guess would be to make the call behave  the same way as in Linux and keeping the name?

Reply:
> Any suggestions for name? Another option I guess would be to make the call behave the same way as in Linux and keeping the name?

I do not have any good suggestion for a name (or actually it would be prefix in this case), but as you said, the best option would be to make the API to work same way as in Linux in order to avoid conflicts.

Reply:
I don't have any suggestion either. I think it'd be better if @andyross commented in this regard as he has a better understanding of what is good to prefix things with.
A benefit of using different names though is that then the issue with the POSIX arch name conflicts are fixed without extra tricks.

Reply:
We could use `sys_` prefix here if nothing better can be figured out.

Reply:
> Another option I guess would be to make the call behave the same way as in Linux and keeping the name?

That's the preferred way, in this in similar cases.

Reply:
>Another option I guess would be to make the call behave the same way as in Linux and keeping the name?

That'd solve this particular issue.

But note the other issue I mentioned below, that also applies to unit tests: As things stand today, you can neither compile this code targeting the POSIX arch, or build unit tests for it.
In general, I'm certainly inclined towards making small concessions in a design to allow for better test-ability. I'd say adding a prefix to these and the header filename is not a too tough one. So I'd ask for you to consider it as it solves both problems in one go.

Reply:
Note that this isn't actually in POSIX but a Linux specific call. So if the goal really is POSIX I don't think it will be possible. If by POSIX, we really want Linux, then it would be possible to run with the `posix_cheats` thing I guess.

Reply:
> then it would be possible to run with the posix_cheats thing I guess.

@tsvehagen Note that `posix_cheats.h` is a header for the "POSIX arch" (meaning the posix part of the name comes from the arch name), to rename symbols out of the way to avoid conflicts with the underlaying libC/OS calls. Not just for the POSIX shim/API, even if that most symbols renamed there are POSIX API related.

Reply:
@tsvehagen:

> Note that this isn't actually in POSIX but a Linux specific call.

Yes, but Linux is an OS based on POSIX principles. It doesn't follow the POSIX x.y.z standard straitjacket though, it goes far and beyond it (so, in *teen years, if Linux did it really well, its functionality might become a new volume of the POSIX standard). Let's call it "extended POSIX", shortened for "posix" in colloquial talk.

The point however is that the functionality proposed here is: a) based on, and intended to interact, with functionality provided by Zephyr POSIX subsystem; b) it's based on Linux functionality, and not Zephyr's NIH. Then the obvious choices would be: a) treat it as part of Zephyr POSIX subsys; b) follow Linux interface as closely as possibly, definitely not introduce avoidable discrepancies (like different function signatures).

This would also automatically address the issue which @aescolar brings: Zephyr POSIX subsys is officially not compatible with "native_posix" Zephyr target (due to the limitations of the latter). This is well-known issue: https://github.com/zephyrproject-rtos/zephyr/issues/13054 , which, when we checked with @aescolar last Novemeber, there was still no interest to resolve (@aescolar, feel free to correct me on exact formulation of the status). 

So, my suggestion would be to move the functionality to `lib/posix/`. Or if there's strong desire to track extensions explicitly, to `lib/posix_ext/`.

I'd also like to bring attention that idea to add eventfd functionality isn't new, we used to have a feature request for that: https://github.com/zephyrproject-rtos/zephyr/issues/16376, which was closed due to lack of motion. Likewise, there's idea to implement epoll() (e.g. https://lists.zephyrproject.org/g/devel/topic/25004178), which also lacks critical mass so far.

So, thanks for coming forward with this, but my suggestion would be to look how to integrate these in/with existing Zephyr frameworks, instead of treating it as a thing in itself.

Reply:
I totally agree. I also read your comment https://github.com/zephyrproject-rtos/zephyr/issues/17706#issuecomment-588881519. Very well written :+1: 

What I meant to point out was just that there might still be some confusion as something that runs on for example `native_posix` on a Linux machine, which has a native eventfd, might not work on macOS for example.

> So, thanks for coming forward with this, but my suggestion would be to look how to integrate these in/with existing Zephyr frameworks, instead of treating it as a thing in itself.

Sounds like a good idea. So to sum up I'll make the signature and behaviour like the eventfd in Linux and move it to `lib/posix`.